R/f7-inputs.R

```diff
@@ -841,40 +849,80 @@ createInputLayout <- function(
   classes <- c(
     "item-input",
     if (outline) "item-input-outline",
-    if (info) "item-input-with-info"
+    if (!is.null(info)) "item-input-with-info"
   )
 
   item <- htmltools::tagQuery(item)$
     find(".item-content")$
-    addClass(classes)$allTags()
+    addClass(classes)$
+    allTags()
 
   # Add label on title + add input wrap sibling
-  item <- htmltools::tagQuery(item)$
-    find(".item-title")$
-    addClass("item-label")$
-    after(
-    shiny::tags$div(
-      class = paste0(
-        "item-input-wrap",
-        if (dropdown) " input-dropdown-wrap"
-      ),
-      ...,
-      if (clear) shiny::span(class = "input-clear-button")
-    )
-  )$allTags()
+  if (!is.null(label)) {
+    item <- htmltools::tagQuery(item)$
+      find(".item-title")$
+      addClass("item-label")$
+      allTags()
+  }
+
+  innerItems <- shiny::tags$div(
+    class = paste0(
+      "item-input-wrap",
+      if (dropdown) " input-dropdown-wrap"
+    ),
+    ...,
+    if (clear) shiny::span(class = "input-clear-button"),
+    if (!is.null(info)) {
+      shiny::tags$div(
+        class = "item-input-info",
+        info
+      )
+    }
+  )
+
+  item <- if (is.null(label)) {
+    htmltools::tagQuery(item)$
+      find(".item-inner")$
+      append(innerItems)$
+      allTags()
+  } else {
+    htmltools::tagQuery(item)$
+      find(".item-title")$
+      after(innerItems)$
+      allTags()
+  }
+
+  # Remove extra item-title-row class if media
+  if (!is.null(media) && !is.null(label)) {
+    inner <- htmltools::tagQuery(item)$
+      find(".item-title-row")$
+      selectedTags()[[1]]$children
+
+    item <- htmltools::tagQuery(item)$
+      find(".item-title-row")$
+      replaceWith(inner)$
+      allTags()
+  }
+
 
   if (floating) {
     item <- htmltools::tagQuery(item)$
       find(".item-title")$
-      addClass("item-floating-label")$allTags()
+      removeClass("item-label")$
+      addClass("item-floating-label")$
+      allTags()
   }
-  f7List(
-    inset = inset,
-    outline = outline,
-    dividers = dividers,
-    strong = strong,
-    item
-  )
+
+  # Is item wrapped in a f7List? If yes, we return it,
+  # if no, we wrap it to avoid rendering issues.
+  is_wrapped <- sum(
+    grepl(
+      "^f7List",
+      perl = TRUE,
+      as.character(sys.calls())
+    )
+  ) == 1
+  if (is_wrapped) item else f7List(item)
```

Comment:
not sure how to expose list parameter here ... if we expose `inset` or `strong`, it is only accessible when the item isn't manually wrapped by `f7List`.

Reply:
But if it is manually wrapped by the user with `f7List`, the user would pass the params there right? So  some scenarios:
1. User used `f7List` -> users specifies styles in `f7List`.
2. User doesn't use `f7List` -> we wrap the item in `f7List`, and in the item, we expose list styles params that get passed to `f7List`.
3. User doesn't understand it and provides item in `f7List` but sets styles in the item: we provide a message saying styles params should be passed to `f7List` instead.

But this is slightly confusing. Can't we expect that if somebody really wants to set `strong`, `inset` and other styles, they should always wrap the item in `f7List`?

Reply:
Yes in the end, I recommend people to wrap inputs in f7List or f7Block so we don't need to handle this styling from the input function. Except for some inputs like slider, which have to be wrapped in a list anyway. Also, to avoid a big breaking change (initially all inputs were wrapped in a list, so the new code would not render well), I introduced a `listify` helper which wraps the input in a list if it is not done by the user, the latter can't be styled but this is fine, the goal is just to render as in older shinyMobile versions.